# Prova módulo 8 - Estatística 2

_____
_____
_____

**1) Julgue a afirmativa a seguir como verdadeira ou falsa, justificando sua resposta se a afirmativa for falsa**

Há situações em que a resolução de um problema de negócio demanda o desenvolvimento de um modelo preditivo. Nestes casos, é comum seguirmos uma abordagem data-driven, comumente apoiada por técnicas de aprendizagem de máquina. 

Neste contexto, podemos dizer que um modelo de machine learning nada mais é do que uma representação matemática do problema sob análise (ou de alguns de seus aspectos), construído teoricamente a partir dos princípios fundamentais que regem este problema.

() Verdadeiro

(X) Falso aboragem data-driven coletamos amostras do universo de interesse para que tiremos conclusões esta abordagem é muito mais dependente da estatística.


__________
__________

**2) Julgue a afirmativa a seguir como verdadeira ou falsa, justificando sua resposta se a afirmativa for falsa**

Utilizamos uma abordagem data-driven, baseada em aprendizagem de máquina, em problemas tais que:

- Existe um padrão a ser descoberto (que assumimos ser o processo teórico $\mathcal{F}$ que rege o problema);
- Uma descrição teórica não é praticável;
- Existam dados coletados sobre o problema, isto é, amostras geradas pelo proceso teórico $\mathcal{F}$.

(X) Verdadeiro

( ) Falso


__________
__________

**3) Julgue a afirmativa a seguir como verdadeira ou falsa, justificando sua resposta se a afirmativa for falsa**

Para aferir a capacidade de generalização de um modelo supervisionado, é muito importante que o volume de dados original seja separado nas amostras de treino e de teste. 

O modelo é treinado utilizando os dados de treino; e os dados de teste são utilizados ao fim do processo, para avaliar sua performance em dados não vistos em treinamento, estimando, assim, sua capacidade de generalização.

(X) Verdadeiro

( ) Falso


__________
__________

**4) Considere a figura a seguir, sobre o famoso tradeoff *viés/variância* (*bias/variance*):**

<img src=https://www.learnopencv.com/wp-content/uploads/2017/02/Bias-Variance-Tradeoff-In-Machine-Learning-1.png width=400>

Sobre este tradeoff, avalie as afirmativas a seguir como verdadeiras ou falsas:

- 1 - Dizemos que um modelo tem **alto viés** (bias), se ele sofre de **underfitting**, isto é, se a diferença entre os erros de treino e teste/validação é pequena, e ambos erros são grandes. Este fenômeno tende a ocorrer quando a hipótese não é capaz de capturar corretamente os padrões refletidos na amostra de treino.

- 2 - Dizemos que um modelo tem **alta variância** (variance), se ele sofre de **overfitting**, isto é, se a diferença entre os erros de treino e teste/validação é muito grande, sendo os erros de teste/validação muito maiores que os erros de treino. Este fenômeno tende a ocorrer quando a hipótese é tão complexa, que se torna capaz de capturar até mesmo as particularidades (ruídos) da base de treino, perdendo, assim, poder de generalização.

Julgando as afirmativas 1 e 2, temos, respectivamente:

(X) Verdadeiro / Verdadeiro

( ) Falso / Verdadeiro

( ) Verdadeiro / Falso

( ) Falso / Falso


__________
__________

**5) Considere a figura a seguir, sobre métricas de avaliação de problemas de classificação:**

<img src=https://upload.wikimedia.org/wikipedia/commons/thumb/2/26/Precisionrecall.svg/1200px-Precisionrecall.svg.png width=200>

Neste contexto, julgue a afirmativa a seguir como verdadeira ou falsa, justificando sua resposta se a afirmativa for falsa:

Em um sistema/teste de detecção de doenças extremamente contagiosas, é muito importante que o modelo consiga capturar corretamente todas as pessoas que estão de fato doentes, mesmo que isso implique em indicar erroneamente que algumas pessoas saudáveis estão doentes. 

Ou seja, em um sistema desses, é importante que os **falsos negativos** sejam minimizados; portanto, o objetivo é **maximizar** a métrica **precision**.

( ) Verdadeiro

(X) Falso
Maximizar Recall minimiza os falsos negativos, Maximizar Precision minimiza os falsos positivos

__________
__________

**6) Julgue a afirmativa a seguir como verdadeira ou falsa, justificando sua resposta se a afirmativa for falsa**

O método de regressão linear é capaz de gerar, unicamente, hipóteses lineares, pelo fato de ser linear nas features. 

Por este motivo, é impossível, por exemplo, que um modelo construído com o método de regressão linear descreva bem um conjunto de dados que são quadráticos, independente do tipo de pré-processamento que for aplicado às features.

( ) Verdadeiro

(X) Falso O modelo é linear nos parâmetros, não nas features. O termo quadrático que incluímos pode ser considerado como uma nova feature linear.


__________
__________

**7) Julgue a afirmativa a seguir como verdadeira ou falsa, justificando sua resposta se a afirmativa for falsa**

No método de regressão logística aplicado à classificação binária, modelamos explicitamente a probabilidade de uma dada observação pertencer a uma das classes (comumente a classe 1), isto é, $P(y=1 \mid \vec{x})$. 

No entanto, a decisão final de classificação é tomada ao escolhermos um threshold (cutoff), que corta a probabilidade modelada em duas regiões, cada uma referente a uma das classes. 

Ex.: caso escolhamos um threshold de 0.5, temos que se $P(y=1 \mid \vec{x}) > 0.5$, atribuímos $\hat{y} = 1$ para a observação $\vec{x}$; caso contrário, $\hat{y}=0$.

(X) Verdadeiro

() Falso 


__________
__________

**8) Julgue as afirmativas a seguir como verdadeiras ou falsas, justificando sua resposta para as afirmativas falsas, se houver alguma**

> - 1 - Com a aplicação de técnicas de regularização, overfitting é evitado a partir da redução do espaço de hipóteses acessível;

> - 2 - A regularização L1 (lasso), pode implementar feature selection, ao efetivamente zerar os parâmetros referentes a determinadas features;

> - 3 - A regularização L2 (ridge), pode implementar feature selection, ao efetivamente zerar os parâmetros referentes a determinadas features;


Julgando as afirmativas 1, 2 e 3, temos, respectivamente:

(X) Verdadeiro / Verdadeiro / Falso Ridge não efetua feature selection, mas pode minimizar feature

( ) Verdadeiro / Verdadeiro / Verdadeiro

( ) Verdadeiro / Falso / Verdadeiro

( ) Falso / Falso / Falso


__________
__________

**9) Considere as associações, e determine quais estão corretas**

O scikit-learn é uma biblioteca que conta com diversas ferramentas que auxiliam em diversas etapas do processo de modelagem: desde a separação dos dados, até o treinamento e avaliação de modelos.

A biblioteca foi construída em uma estrutura bem unificada, organizada e modularizada. Por exemplo, é comum que determinadas ferramentas estejam em submódulos específicos! 

Sobre isso, considere a associação a seguir entre classes/funções e submódulos (considere a versão mais recente do sklearn!):

- 1 - `LogisticRegression`, `LinearRegression`, `ElasticNet`, estão no submódulo `sklearn.linear_model` <br><br>

- 2 - `PolynomialFeatures`, `StandardScaler`, `MinMaxScaler` estão no submódulo `sklearn.preprocessing` <br><br>

- 3 - `f1_score`, `classification_report`, `confusion_matrix`, estão no submódulo `sklearn.metrics` <br><br>

- 4 - `PolynomialFeatures`, `train_test_split`, `MinMaxScaler` estão no submódulo `sklearn.preprocessing` <br><br>

Assinale a alternativa que contém **apenas associações corretas**:

( ) 1, 2, 3, 4

( ) 2, 4

(X) 1, 2, 3

( ) 1, 3, 4


__________
__________

**10) Vamos treinar e avaliar um modelo!**

Considere o dataset a seguir (**use exatamente o mesmo código!**):


In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split

X, y = make_regression(n_samples = 400,
                       n_features = 5, n_informative = 3, 
                       noise = 25,
                       random_state = 42)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [2]:
from sklearn.linear_model import LinearRegression

In [3]:
reglin = LinearRegression()

reglin.fit(X_train, y_train)

LinearRegression()

In [4]:
y_pred = reglin.predict(X_test)

In [5]:
y_pred_train = reglin.predict(X_train)

In [6]:
from sklearn.metrics import mean_absolute_error

In [7]:
mae_test = mean_absolute_error(y_test, y_pred)

In [8]:
mae_train = mean_absolute_error(y_train, y_pred_train)

In [9]:
mae_train

18.981822066557715

In [10]:
mae_test

19.861718146310462

In [11]:
mae_train - mae_test

-0.8798960797527471

In [12]:
mae_test - mae_train

0.8798960797527471

Ao treinar e avaliar um modelo de regressão linear (todos os hiperparâmetros default), qual é a diferença entra o **erro absoluto médio (MAE)**, calculado nas bases de teste e treino?

( ) 2.56

(X) 0.88

( ) Impossível de calcular esta métrica para os dados disponibilizados

( ) 0

( ) -3.14